lib/dialects/postgres/connection-manager.js

```diff
@@ -18,40 +18,67 @@ class ConnectionManager extends AbstractConnectionManager {
     const pgLib = this._loadDialectModule('pg');
     this.lib = this.sequelize.config.native ? pgLib.native : pgLib;
 
+    this._clearDynamicOIDs();
     this._clearTypeParser();
     this.refreshTypeParser(dataTypes.postgres);
   }
 
   // Expose this as a method so that the parsing may be updated when the user has added additional, custom types
   _refreshTypeParser(dataType) {
-    if (dataType.types.postgres.oids) {
-      for (const oid of dataType.types.postgres.oids) {
-        this.oidMap[oid] = value => dataType.parse(value, oid, this.lib.types.getTypeParser);
+    const arrayParserBuilder = parser => {
+      return value => this.lib.types.arrayParser.create(value, parser).parse();
+    };
+    const rangeParserBuilder = parser => {
+      return value => dataType.parse(value, {parser});
+    };
+
+    // Set range parsers
+    if (dataType.key.toLowerCase() === 'range') {
+      for (const name in this.nameOidMap) {
+        const entry = this.nameOidMap[name];
+        if (! entry.rangeOid) continue;
+
+        const rangeParser = rangeParserBuilder(this.getTypeParser(entry.oid));
+        const arrayRangeParser = arrayParserBuilder(rangeParser);
+
+        this.oidParserMap.set(entry.rangeOid, rangeParser);
+        if (! entry.arrayRangeOid) continue;
+        this.oidParserMap.set(entry.arrayRangeOid, arrayRangeParser);
       }
+      return;
     }
 
-    if (dataType.types.postgres.array_oids) {
-      for (const oid of dataType.types.postgres.array_oids) {
-        this.arrayOidMap[oid] = value => {
-          return this.lib.types.arrayParser.create(value, v =>
-            dataType.parse(v, oid, this.lib.types.getTypeParser)
-          ).parse();
-        };
-      }
+    // Create parsers for normal or enum data types
+    const parser = value => dataType.parse(value);
+    const arrayParser = arrayParserBuilder(parser);
+
+    // Set enum parsers
+    if (dataType.key.toLowerCase() === 'enum') {
+      this.enumOids.oids.forEach(oid => {
+        this.oidParserMap.set(oid, parser);
+      });
+      this.enumOids.arrayOids.forEach(arrayOid => {
+        this.oidParserMap.set(arrayOid, arrayParser);
+      });
+      return;
     }
+
+    // Set parsers for normal data types
+    dataType.types.postgres.forEach(name => {
+      if (! this.nameOidMap[name]) return;
+      this.oidParserMap.set(this.nameOidMap[name].oid, parser);
+
+      if (! this.nameOidMap[name].arrayOid) return;
+      this.oidParserMap.set(this.nameOidMap[name].arrayOid, arrayParser);
+    });
   }
 
   _clearTypeParser() {
-    this.oidMap = {};
-    this.arrayOidMap = {};
+    this.oidParserMap = new Map();
```

Comment:
You are using a `Map`, we actually have an internal support for such type mapping. I assume in a single system, all type oids are unique. Please have a look at https://github.com/sequelize/sequelize/blob/master/lib/dialects/mysql/connection-manager.js#L10 parserStore implementation form other dialects. We can use that for postgres as well.

Reply:
Thanks for the suggestion. At the beginning I was excited about this, I thought this could save some code. Unfortunately, this is what I have found out after implement it.

  - Net increase of 20-30 lines of code
  - Would require sub-stores under each dialect in `parserStore`
  - Would require to map each sub-store to a connection manager
      - In my experiment I used the connection URL to map them
  - Would require to create a new postgres-specific method or the ability to handle postgres specificities
      - The builtin `.set()` method of `Map` is enough

In postgres, each `connectionManager` needs a separate store. The reason is that OIDs can be mapped to different data types across postgres instances. `parserStore` is a centralized database. To make `parserStore` compatible, it should allow multiple sub-stores for each dialect. These sub-stores need to be mapped to each connection manager, which increases complexity.

The `.refresh(`) method used by parserStore is not the one that postgres needs. `.refresh()` assumes that `dataType.parser` is the parser and that the names `dataType.types.postgres` are the keys to index them. Postgres is different, it uses OIDs instead of names to find the parser. On top of that, composite dataTypes(arrays and ranges) need special parsers. I firmly believe that the `.set(<key/oid>, <value/parser>)` method is the most appropiate.